libbeat/esleg/eslegclient/connection.go

```diff
@@ -0,0 +1,435 @@
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/transport"
+	"github.com/elastic/beats/libbeat/common/transport/tlscommon"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/testing"
+)
+
+// Connection manages the connection for a given client.
+type Connection struct {
+	ConnectionSettings
+
+	Encoder BodyEncoder
+	HTTP    *http.Client
+
+	version common.Version
+	log     *logp.Logger
+}
+
+// ConnectionSettings are the settings needed for a Connection
+type ConnectionSettings struct {
+	URL          string
+	Proxy        *url.URL
+	ProxyDisable bool
+
+	Username string
+	Password string
+	APIKey   string
+	Headers  map[string]string
+
+	TLS *tlscommon.TLSConfig
+
+	OnConnectCallback func() error
+	Observer          transport.IOStatser
+
+	Parameters       map[string]string
+	CompressionLevel int
+	EscapeHTML       bool
+	Timeout          time.Duration
+}
+
+// NewConnection returns a new Elasticsearch client
+func NewConnection(s ConnectionSettings) (*Connection, error) {
+	u, err := url.Parse(s.URL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse elasticsearch URL: %v", err)
+	}
+
+	if u.User != nil {
+		s.Username = u.User.Username()
+		s.Password, _ = u.User.Password()
+		u.User = nil
+
+		// Re-write URL without credentials.
+		s.URL = u.String()
+	}
+	logp.Info("elasticsearch url: %s", s.URL)
+
+	// TODO: add socks5 proxy support
+	var dialer, tlsDialer transport.Dialer
+
+	dialer = transport.NetDialer(s.Timeout)
+	tlsDialer, err = transport.TLSDialer(dialer, s.TLS, s.Timeout)
+	if err != nil {
+		return nil, err
+	}
+
+	if st := s.Observer; st != nil {
+		dialer = transport.StatsDialer(dialer, st)
+		tlsDialer = transport.StatsDialer(tlsDialer, st)
+	}
+
+	var encoder BodyEncoder
+	compression := s.CompressionLevel
+	if compression == 0 {
+		encoder = NewJSONEncoder(nil, s.EscapeHTML)
+	} else {
+		encoder, err = NewGzipEncoder(compression, nil, s.EscapeHTML)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var proxy func(*http.Request) (*url.URL, error)
+	if !s.ProxyDisable {
+		proxy = http.ProxyFromEnvironment
+		if s.Proxy != nil {
+			proxy = http.ProxyURL(s.Proxy)
+		}
+	}
+
+	return &Connection{
+		ConnectionSettings: s,
+		HTTP: &http.Client{
+			Transport: &http.Transport{
+				Dial:            dialer.Dial,
+				DialTLS:         tlsDialer.Dial,
+				TLSClientConfig: s.TLS.ToConfig(),
+				Proxy:           proxy,
+			},
+			Timeout: s.Timeout,
+		},
+		Encoder: encoder,
+		log:     logp.NewLogger("esclientleg"),
+	}, nil
+}
+
+// NewClients returns a list of Elasticsearch clients based on the given
+// configuration. It accepts the same configuration parameters as the Elasticsearch
+// output, except for the output specific configuration options.  If multiple hosts
+// are defined in the configuration, a client is returned for each of them.
+func NewClients(cfg *common.Config) ([]Connection, error) {
```

Comment:
Do we need this outside of the Elasticsearch output?

Reply:
Actually, this is not called from the Elasticsearch output at all. Within the Beats codebase it is only called from `eslegclient.NewConnectedClient` (which is defined in the same file as this function). However, APM server is calling this as well. See https://github.com/elastic/apm-server/blob/72c446d4cb7904ec27198a887f09d6f7764aad2a/beater/beater.go#L210.